src/VecSim/algorithms/brute_force/bf_batch_iterator.cpp

```diff
@@ -105,23 +107,45 @@ VecSimQueryResult *BF_BatchIterator::selectBasedSearch(size_t n_res) {
     // will be placed before it, and all the rest will be placed after.
     std::nth_element(valid_begin_it, n_th_element_pos, this->scores.end());
 
-    auto *results = array_new<VecSimQueryResult>(n_res);
+    rl.results = array_new<VecSimQueryResult>(n_res);
     for (size_t i = this->scores_valid_start_pos; i < this->scores_valid_start_pos + n_res; i++) {
-        results = array_append(results, VecSimQueryResult{});
-        VecSimQueryResult_SetId(results[array_len(results) - 1], this->scores[i].second);
-        VecSimQueryResult_SetScore(results[array_len(results) - 1], this->scores[i].first);
+        rl.results = array_append(rl.results, VecSimQueryResult{});
+        VecSimQueryResult_SetId(rl.results[array_len(rl.results) - 1], this->scores[i].second);
+        VecSimQueryResult_SetScore(rl.results[array_len(rl.results) - 1], this->scores[i].first);
     }
     // Update the valid results start position after returning the results.
-    this->scores_valid_start_pos += array_len(results);
-    return results;
+    this->scores_valid_start_pos += array_len(rl.results);
+    return rl;
 }
 
 BF_BatchIterator::BF_BatchIterator(void *query_vector, const BruteForceIndex *bf_index,
+                                   VecSimQueryParams *queryParams,
                                    std::shared_ptr<VecSimAllocator> allocator)
-    : VecSimBatchIterator(query_vector, allocator), index(bf_index), scores_valid_start_pos(0) {
+    : VecSimBatchIterator(query_vector, queryParams ? queryParams->timeoutCtx : nullptr, allocator),
+      index(bf_index), scores_valid_start_pos(0) {
     BF_BatchIterator::next_id++;
 }
 
+// Compute the score for every vector in the block by using the given distance function.
+// Return a collection of (score, label) pairs for every vector in the block.
+vecsim_stl::vector<std::pair<float, labelType>>
+BF_BatchIterator::computeBlockScores(VectorBlock *block, VecSimQueryResult_Code *rc) {
+    const void *queryBlob = getQueryBlob();
+    DISTFUNC<float> DistFunc = getIndex()->distFunc();
+    size_t len = block->getLength();
+    vecsim_stl::vector<std::pair<float, labelType>> scores(len, this->allocator);
+    for (size_t i = 0; i < len; i++) {
+        if (__builtin_expect(VecSimIndex::timeoutCallback(this->getTimeoutCtx()), 0)) {
+            *rc = VecSim_QueryResult_TimedOut;
+            return vecsim_stl::vector<std::pair<float, labelType>>(this->allocator);
```

Comment:
can we have a static empty list to return each time instead of creating a new one?

Reply:
I will return the current result instead, and the caller will decide what to do with it (aka ignore)
is this better?